File: python/noc/simulate.py
import numpy as np
import math
from scipy.integrate import odeint

def impulse(a, tau, n, inp, duration, c=None):
    def f(y, t):
        dx = y.dot(a.T) / tau
        dx = dx + (inp / tau) if t > 0.05 and t < 0.054 else dx
        return dx

    dt = 1e-3
    n_steps = int(duration / dt) + 1
    ts = np.linspace(0, duration, n_steps)
    xs = odeint(f, np.zeros(n), ts, tcrit=np.linspace(0.04, 0.06, 20))
    return (ts, xs) if c is None else (ts, xs.dot(c.T))

def ou_process(n, dt, tau, duration): 
    t_max = 3 + int(duration / dt)   
    table = np.zeros((t_max, n))
    x = np.random.randn(1,n)
    for i in range(t_max): 
        x = x if (i==0) else ((1 - (dt / tau))* x) + (math.sqrt(2 * dt / tau) * np.random.randn(1,n))
        table[i,:] = x
    return (table, lambda t : table[int(t/dt),:])

File: ocaml/noc/simulate.ml
open Owl
open Owl_ode
open Owl_ode.Types

let impulse ?c ~tau ~n ~a ~input duration =
  let tspec = T1 { t0 = 0.0; dt = 1E-3; duration } in
  let f =
    let z = Mat.(transpose a /$ tau) in
    fun x t ->
      let dx = Mat.(x *@ z) in
      if t < 0.05 || t > 0.055 then dx else Mat.(dx + (input /$ tau))
  in
  let ts, xs = Ode.odeint (module Owl_ode.Native.D.RK4) f Mat.(zeros 1 n) tspec () in
  let xs =
    match c with
    | Some c -> Mat.(xs *@ transpose c)
    | None   -> xs
  in
  Mat.concat_horizontal ts xs


let ou_process ~n ~dt ~tau duration =
  let t_max = 1 + int_of_float (duration /. dt) in
  let rec iter t x accu =
    if t = t_max
    then accu |> List.rev |> Array.of_list |> Mat.concatenate ~axis:0
    else (
      let accu = x :: accu in
      let x =
        Mat.((1. -. (dt /. tau) $* x) + (Maths.(sqrt (2. *. dt /. tau)) $* gaussian 1 n))
      in
      iter (succ t) x accu)
  in
  let table = iter 0 Mat.(gaussian 1 n) [] in
  table, fun t -> Mat.row table (int_of_float (t /. dt))
